feat(compiler): Wire offline compilation as a build step.

alexeagle opened the pull request:
cc @gdi2290

alexeagle commented on tools/metadata/src/collector.ts:
```diff
@@ -48,16 +48,20 @@ function moduleNameFromBaseName(moduleFileName: string, baseFileName: string): s
  */
 export class MetadataCollector {
-  constructor() {}
+  constructor(private resolveModuleFromPath: (path: string) => string = (p) => null) {}
```
@chuckjaz here is where this function signature could go into a host interface. I suppose I would just add that interface declaration in this file, maybe before the MetadataCollector class?

chuckjaz replied:
Something like,
```typescript
export interface MetadataCollectorHost {
  resolveModuleFromPath?(path: string): string;
}
```

That way we could add another method without breaking the API.

robwormald commented on gulpfile.js:
```diff
@@ -452,22 +452,40 @@ gulp.task('serve.e2e.dart', ['build.js.cjs'], function(neverDone) {
+  execProcess('tsc', '-p ' + project, function(e, stdout, stderr) {
+    if (e) {
+      console.log(stdout);
+      console.err(stderr);
```
console.err => console.error

robwormald commented on tools/compiler_cli/index.ts:
```diff
@@ -0,0 +1,110 @@
+  const program = ts.createProgram(parsed.fileNames, parsed.options, host);
+  const collector = new MetadataCollector((fileName: string) => filenameToModule[fileName]);
+
+  const host: LinkerHost = {
```
redefining host from line 59

robwormald commented:
i know this is still WIP, so sorry for the comments, but one bit of weirdness:
```
import {Component} from 'angular2/core' 

@Component({ ... })
export SomeClass {}
```
doesn't work, and throws (after some debugging) with a `'No Directive annotation found on SomeClass',` error. 

using 
```
import {Component} from 'angular2/src/core/metadata'
```
and doing the same thing works and successfully generates a .template file. The test code here does the same thing (took me a while to recognize that), is that expected? Some artifact of recent changes?

mhevery commented on tools/compiler_cli/test/basic.html:
```diff
@@ -0,0 +1 @@
+<div Igor>{{ctxProp}}</div>
```
why `Igor`? What does this flag do?

mhevery commented on tools/compiler_cli/test/directives_and_pipes.ts:
```diff
@@ -0,0 +1,12 @@
+@Component({
+  selector: 'comp',
+  template: '<blink *ngIf="true"></blink>',
+  directives: [COMMON_DIRECTIVES],
```
I don't think `COMMON_DIRECTIVES` are required. They should be auto included with `bootstrap`

mhevery commented:
Very exciting!!!

alexeagle commented:
@chuckjaz Rob has a good point above, that I reverse-map based on the path where a module was discovered, to workaround path mapping issues. Any import from a re-export has this problem, because the string `./src/core/metadata` is the location where Component is loaded in angular2/core.ts.

It won't be a problem if you use the collector.ts from head, and angular2 comes from regular node moduleResolution with no pathMapping. That's how users would use it. Just a problem in our test case.

We need a real chat about #8082 ...

chuckjaz commented:
We should go over this tomorrow.

tbosch commented on gulpfile.js:
```diff
@@ -452,22 +452,40 @@ gulp.task('serve.e2e.dart', ['build.js.cjs'], function(neverDone) {
   });
 }
 
+// Gulp-typescript doesn't work with typescript@next:
```
Does your PR still require typescript@next? If not, maybe use gulp-typescript?

alexeagle replied:
Yes, I still use typescript@next to compile the compiler (it uses baseUrl/paths to resolve the 'angular2' imports)

alexeagle replied:
But, the issue referenced here is resolved, so it may be possible to use gulp-typescript instead.
(Not that it adds anything, I want the `tsconfig.json

alexeagle replied:
Tried newer gulp-typescript, it is still broken. Filed a new issue and referenced in this comment.

Note that I think we should stop using gulp-typescript and store our configuration in `tsconfig.json` instead. Maybe in a followup PR I'll remove it :)

tbosch commented on gulpfile.js:
```diff
@@ -1026,6 +1044,31 @@ gulp.task('!test.typings',
+
+gulp.task('!test.compiler_cli.codegen', function(done) {
+  try {
+    require('./dist/js/cjs/compiler_cli')
```
This will prevent multiple calls to pick up changes.
You need to spawn a new node process each time you run the test (see the jasmine tests in https://github.com/angular/angular/blob/master/gulpfile.js#L108)

alexeagle replied:
I don't think I care about calling this multiple times in a single process. There is no watch mode for this end-to-end test. I'd rather keep this simple without `fork` unless there is a good reason?

tbosch commented on gulpfile.js:
```diff
@@ -1026,6 +1044,31 @@ gulp.task('!test.typings',
+  }
+});
+gulp.task('!test.compiler_cli.jasmine', function(done) {
+  // TODO(alexeagle): add some jasmine tests that the app still works after compilation
```
Resolve this TODO

alexeagle replied:
Will change to a TODO to wire up example tests

tbosch commented on gulpfile.js:
```diff
@@ -1026,6 +1044,31 @@ gulp.task('!test.typings',
+    done(err);
+  }
+});
+gulp.task('!test.compiler_cli.jasmine', function(done) {
```
Rename into `!test.unit.compile_cli`

tbosch commented on gulpfile.js:
```diff
@@ -1026,6 +1044,31 @@ gulp.task('!test.typings',
+gulp.task('!build.compiler_cli', ['build.js.cjs'],
+          function(done) { runTsc('tools/compiler_cli/src', done); });
+
+gulp.task('!test.compiler_cli.codegen', function(done) {
```
What is the difference between this task and `!test.compiler_cli.jasmine`?

alexeagle replied:
removed

tbosch commented on tools/compiler_cli/README.md:
```diff
@@ -0,0 +1,72 @@
+Angular applications are built with templates, which may be `.html` or `.css` files,
+or may be inline `template` attributes on Decorators like `@Component`.
+
+These templates are always compiled into executable JS when the application runs.
```
Technically not true as we have an `interpretation` mode as well.

tbosch commented on tools/compiler_cli/README.md:
```diff
@@ -0,0 +1,72 @@
+or may be inline `template` attributes on Decorators like `@Component`.
+
+These templates are always compiled into executable JS when the application runs.
+This compilation can occur on the client, but it results in slower load time, and also
```
Maybe use "slower bootstrap time", as "load time" might be associated to bigger file size which the next part of the sentence is about.

tbosch commented on tools/compiler_cli/README.md:
```diff
@@ -0,0 +1,72 @@
+so set `genDir` to the same location as your files (typicially the same as `rootDir`).
+Add `**/*.ngfactory.ts` to your `.gitignore` or other mechanism for your version control system.
+
+TypeScript 1.9 and above, you can add a generated folder into your application,
```
In TypeScript...

tbosch commented on tools/compiler_cli/src/codegen.ts:
```diff
@@ -0,0 +1,157 @@
+import {wrapCompilerHost} from './compiler_host';
+
+import * as compiler from 'angular2/src/compiler/compiler';
+// TODO(alexeagle): expose these through angular2/src/compiler/compiler as well?
```
Agree. Could you fix this?

alexeagle replied:
done

tbosch commented on tools/compiler_cli/src/codegen.ts:
```diff
@@ -0,0 +1,157 @@
+  // Absolute path to a directory where generated file structure is written
+  genDir: string;
+
+  // TODO(alexeagle): add more options, eg.
```
Remove this TODO for now.

tbosch commented on tools/compiler_cli/src/codegen.ts:
```diff
@@ -0,0 +1,157 @@
+          console.error(e);
+          throw e;
+        }
+        if (!directive.isComponent) {
```
Move the above code into a private method `readComponents(absSourcePath:string):CompileDirectiveMetadata[]`

tbosch commented on tools/compiler_cli/src/codegen.ts:
```diff
@@ -0,0 +1,157 @@
+      return Promise.all(componentMetadatas)
+          .then((metadatas: compiler.CompileDirectiveMetadata[]) => {
+            if (!metadatas || !metadatas.length) {
+              console.log(`Not writing template for ${absSourcePath}: no metadatas`);
```
Should this really be an error / warning? Users might use globs to collect directives so this case is very likely...

alexeagle replied:
agree, I added this for debugging at one point.

tbosch commented on tools/compiler_cli/src/codegen.ts:
```diff
@@ -0,0 +1,157 @@
+            }
+            const generated = this.generateSource(metadatas);
+            const sourceFile = this.program.getSourceFile(absSourcePath);
+            // We use program.emit to map source file locations to output file locations.
```
Add a bit more details here, maybe a concrete example that shows why this is needed.

tbosch replied:
Let's change this: Don't rely on how TS chooses to lay out his output folder, but rather calculate the final gendir based on `rootDirs`. By this, TS won't apply his dist file layout algorithm 2x.

tbosch commented on tools/compiler_cli/src/codegen.ts:
```diff
@@ -0,0 +1,157 @@
+        try {
+          directive = this.resolver.getDirectiveMetadata(<any>staticType);
+        } catch (e) {
+          // TODO: directive resolver throws on non-directive symbols
```
Maybe introduce a method in the resolver that does not throw but returns null?

tbosch replied:
Resolve the TODO

tbosch commented on tools/compiler_cli/src/codegen.ts:
```diff
@@ -0,0 +1,157 @@
+    return Promise.all(promises).then(() => {});
+  }
+
+  // TODO: use DI to create this object graph??
```
Remove this TODO, let's not use DI for now here.

tbosch commented on tools/compiler_cli/src/compiler_host.ts:
```diff
@@ -0,0 +1,90 @@
+  getCanonicalFileName = (fileName: string) => this.delegate.getCanonicalFileName(fileName);
+  useCaseSensitiveFileNames = () => this.delegate.useCaseSensitiveFileNames();
+  getNewLine = () => this.delegate.getNewLine();
+  resolveModuleNames(moduleNames: string[], containingFile: string) {
```
Use arrow function as well.

tbosch commented on tools/compiler_cli/src/compiler_host.ts:
```diff
@@ -0,0 +1,90 @@
+    return this.delegate.resolveModuleNames(moduleNames, containingFile);
+  }
+
+  // TODO(alexeagle): need to express the optionality of this member.
```
Resolve TODO

tbosch commented on tools/compiler_cli/src/compiler_host.ts:
```diff
@@ -0,0 +1,90 @@
+  };
+
+  reverseModuleResolution(fileName: string) {
+    // NB. TS parsed options gives baseUrl relative to basePath, not the same as in tsconfig.
```
This comment is a little cryptic...

tbosch commented on tools/compiler_cli/src/compiler_host.ts:
```diff
@@ -0,0 +1,90 @@
+    return result;
+  };
+
+  reverseModuleResolution(fileName: string) {
```
This method needs to get 2 arguments:
- `containingFile: string`
- `importedFile: string`

MetadataCollectorHost needs to be changed to provide the `containingFile` in addition.

This `ReverseModuleResolutionHost` should be part of the static reflector package, as everyone will need it

tbosch commented on tools/compiler_cli/src/compiler_host.ts:
```diff
@@ -0,0 +1,90 @@
+  }
+}
+
+// TODO(alexeagle): workaround https://github.com/Microsoft/TypeScript/issues/8245 ??
```
Resolve the TODO

tbosch commented on tools/compiler_cli/src/index.ts:
```diff
@@ -0,0 +1,52 @@
+// TODO(alexeagle): use --lib=node when available; remove this reference
```
When will this be available? Tracking bug in TS?

tbosch commented on tools/compiler_cli/src/index.ts:
```diff
@@ -0,0 +1,52 @@
+// TODO(alexeagle): use --lib=node when available; remove this reference
+/// <reference path="../../typings/node/node.d.ts"/>
+
+// Must be imported first.
```
Why?

alexeagle replied:
importing decorators.ts from angular2 still reads reflect metadata and throws at top-level.

Error: reflect-metadata shim is required when using class decorators

tbosch commented on tools/compiler_cli/src/tsc.ts:
```diff
@@ -0,0 +1,104 @@
+/**
+ * Our interface to the TypeScript standard compiler.
+ * If you write an Angular compiler plugin for another build tool,
+ * you should implementa similar interface.
```
Typo: "implementa"

tbosch commented on tools/compiler_cli/src/tsc.ts:
```diff
@@ -0,0 +1,104 @@
+  }
+}
+
+export class TSC implements CompilerInterface {
```
Change name to `Tsc`

tbosch commented on tools/compiler_cli/src/tsc.ts:
```diff
@@ -0,0 +1,104 @@
+
+    check(this.parsed.errors);
+
+    // Default codegen goes to the output directory
```
Should this really be the default as it won't be compiled then...
Make the `genDir` required.

alexeagle replied:
made `genDir` default to `basePath`

tbosch commented on tools/compiler_cli/src/tsconfig.json:
```diff
@@ -0,0 +1,20 @@
+        },
+        "experimentalDecorators": true,
+        "rootDir": ".",
+        // Write to a directory that has the node_modules symlink in a parent
```
Json files don't allow comments

tbosch commented on tools/compiler_cli/src/compiler_host.ts:
```diff
@@ -0,0 +1,90 @@
+        if (moduleName.indexOf(".") !== 0) {
+          const modulePath = path.relative('.', resolved.resolvedModule.resolvedFileName);
+          debug(`Adding reverse mapping ${modulePath} => ${moduleName}`);
+          this.reverseMap[modulePath] = moduleName;
```
Throw if a module can be resolved in 2 different ways.

alexeagle commented:
@tbosch TAL?

chuckjaz committed:
fix(8223): Preserve Provider expressions

Preserves constructor calls in addition to function calls.
Introduced a special case for forwardRef() similar to CONST_EXPR.

alexeagle committed:
fix(codegen): event handler has boolean return type

alexeagle committed:
fix(metadata): expose Providers in metadata

These worked in Dart because they were effectively exported even without the export keyword.
Without exporting these symbols, they are not produced in .metadata.json files, which leaves
dangling references from the Decorators that use them.

alexeagle committed:
chore(compiler): refactoring for offline compiler cli

- pass a baseUrl for asset resolution from static symbols
- fixes in StaticReflector to work with a path-aware host

see #7483

alexeagle committed:
chore(compiler): add dependency on tsickle

This tool lets us re-write TypeScript sources before entering the emit pipeline.
For example, we lower Decorators to the tree-shakable Annotation form.

alexeagle committed:
feat(offline compiler): a replacement for tsc that compiles templates

see #7483.

alexeagle committed:
chore(build): fix formatting and tests

alexeagle commented:
Cleaned up and ready for review

mary-poppins commented:
Merging PR #8098 on behalf of @alexeagle to branch [presubmit-alexeagle-pr-8098](https://github.com/angular/angular/tree/presubmit-alexeagle-pr-8098).
